### PR TITLE
add "show live tabs only" option

### PR DESCRIPTION
**Pull request by KogasaPls**
Pull request checklist:

- [ ] `CHANGELOG.md` was updated, if applicable

# Description
This PR adds a "show live tabs only" option that hides tabs which are not marked as live. By default, it is disabled and has no hotkey bound to it. While active, it locks the tab layout, since it's not clear how a change in layout should translate when offline tabs are unhidden. When tabs are marked as live/not live, they will automatically be inserted/hidden appropriately. 

A video demonstration of the core functionality can be found [here](https://streamable.com/8v900l).

I'd appreciate your feedback. Thank you.

### CODE REVIEW

File: src/controllers/hotkeys/ActionNames.hpp [hunk 1]
```diff
@@ -202,6 +202,13 @@ inline const std::map<HotkeyCategory, ActionDefinitionMap> actionNames{
               "[on, off, or toggle. default: toggle]",
               0,
               1,
+          }},
+         {"setShowLiveOnly",
+          ActionDefinition{
+              "Show live tabs only",
```
Comment (jammehcow) [inline src/controllers/hotkeys/ActionNames.hpp:208]: Wording was a bit backwards
```suggestion
              "Only show live tabs",
```

File: src/widgets/Notebook.cpp [hunk 2]
```diff
@@ -390,6 +404,21 @@ void Notebook::setShowTabs(bool value)
     }
 }
 
+bool Notebook::getShowLiveOnly() const
+{
+    return this->showLiveOnly_;
+}
+void Notebook::setShowLiveOnly(bool value)
```
Comment (jammehcow) [inline src/widgets/Notebook.cpp:411]: nitpick: formatting
```suggestion
}

void Notebook::setShowLiveOnly(bool value)
```

File: src/widgets/Notebook.cpp [hunk 3]
```diff
@@ -390,6 +404,21 @@ void Notebook::setShowTabs(bool value)
...
+{
+    this->showLiveOnly_ = value;
+    this->showLiveOnlyAction_->setChecked(value);
+    //check and disable the "lock tabs" button
+    this->lockNotebookLayoutAction_->setEnabled(!value);
```
Comment (jammehcow) [inline src/widgets/Notebook.cpp:416]: nitpick: formatting
```suggestion
    this->showLiveOnlyAction_->setChecked(value);

    // check and disable the "lock tabs" button
    this->lockNotebookLayoutAction_->setEnabled(!value);
```

File: src/widgets/Notebook.hpp [hunk 4]
```diff
@@ -80,6 +83,10 @@ class Notebook : public BaseWidget
         QWidget *selectedWidget{};
     };
 
+    const QList<Item> allItems()
+    {
+        return allItems_;
+    }
     const QList<Item> items()
```
Comment (jammehcow) [inline src/widgets/Notebook.hpp:90]: nitpick: formatting
```suggestion
    }

    const QList<Item> items()
```

File: src/widgets/Window.cpp [hunk 5]
```diff
@@ -660,6 +660,53 @@ void Window::addShortcuts()
...
+                 getSettings()->showLiveOnly.setValue(true);
+                 this->notebook_->setShowLiveOnly(true);
+             }
+             else if (mode == 2)
+             {
+                 auto value = this->notebook_->getShowLiveOnly();
+                 getSettings()->showLiveOnly.setValue(!value);
+                 this->notebook_->setShowLiveOnly(!value);
+             }
+             return "";
```
Comment (jammehcow) [inline src/widgets/Window.cpp:708]: You've essentially duplicated your if/else statements here. I'd recommend that you move all of your `mode` conditional logic into the first lot of statements and remove the `mode` variable altogether.

**Review by jammehcow — CHANGES_REQUESTED** [review]
Other than the changes I've suggested as part of the review, I've tested and found two bugs:

1. Toggling this feature and opening a new tab causes an empty spit to appear in the first position (on top of settings/accounts buttons + the first tab):

https://user-images.githubusercontent.com/4962764/162606396-07d2b6e2-16ef-4edb-aaaf-bdb908b835c9.mp4

2. Ctrl + Tab allows you to navigate all tabs even when they're hidden, but the quick switcher (Ctrl + K) does not switch to the hidden channels.

https://user-images.githubusercontent.com/4962764/162606401-087b0040-8e37-4e3f-b316-fd37360a9412.mp4

Nice work so far but has a few kinks that need ironing out.

**Comment by pajlada** [comment]
Super interesting feature! Would love to get this merged in. Haven't had a time to review the actual code yet but thank you for working on it!

File: src/widgets/Window.cpp (hunk 5 above)
Comment (KogasaPls) [reply src/widgets/Window.cpp]: Thanks for the review! I simplified this bit in 51ac47b.

**Comment by KogasaPls** [comment]
I've addressed both bugs in this last set of changes.

CTRL+(SHIFT)+Tab will now cycle through only the _displayed_ tabs, and the channel switcher will now correctly focus offline tabs. Would it be better for the channel switcher filter out offline tabs instead?

**Comment by KogasaPls** [comment]
I've noticed that I can't change the channel of a split while the setting is enabled. I'll look into fixing that.

**Comment by zneix** [comment]
I don't like the idea this "feature" brings, wouldn't this clash with "toggle visibility of tabs"? shouldn't rather already existing tab visibility 0(hidden)/1(visible) be changed to something like an enum? e.g. 0/1/2 - hidden, all visible, only live visible, etc... if someone decides to add more things and maintain better future-compatibility.

**Comment by jammehcow** [comment]
Agree with @zneix on this one, combining the two similar features into a single setting would make this a whole lot nicer to work with in the future - should we need to add something else around tab visibility.

**Comment by maxcamorra** [comment]
Will I stay in the chat of the offline streamer as well or will it automatically leave these and only stay inside the live streamer?

**Comment by leon-richardt** [comment]
> Will I stay in the chat of the offline streamer as well or will it automatically leave these and only stay inside the live streamer?

Toggling this setting yields a strictly visual change. You will stay connected to offline chat rooms in the background, only their tabs will not be shown in the notebook.

**Review by pajlada — CHANGES_REQUESTED** [review]
Small questions/tweaks, then I'll do a re-review

**Comment by pajlada** [comment]
Closing in favour of https://github.com/Chatterino/chatterino2/pull/4358

**Pull request closed without merging**